app/controllers/institutions_controller.rb

```diff
@@ -2,43 +2,67 @@
 
 # Controller for reading and filering the school
 class InstitutionsController < ApplicationController
-  before_action :ensure_page_size_is_positive
-
+  before_action :extract_page_presence, :ensure_page_size_is_positive, :area_query, :name_query, :page_size,
+                :public_school, :school_rspo_type_ids, only: [:index]
+  before_action :institution_id, only: [:show]
   def index
-    page = extract_page_presence
-    institutions = Institution.all.paginate(page: page, per_page: @page_size)
-    render status: '200', json: institutions
+    render status: :bad_request, json: { message: @error_message } and return unless @error_message.nil?
+
+    institutions = Institution.all
+    institutions = institutions.search_by_area(@area_query) unless @area_query.nil?
+    institutions = institutions.search_by_name(@name_query) unless @name_query.nil?
+    institutions = institutions.where(rspo_institution_type_id: @school_rspo_type_ids) unless @school_rspo_type_ids.nil?
+    institutions = institutions.where(public: @public_school) unless @public_school.nil?
+
+    render status: '200', json: institutions.paginate(page: @page, per_page: @page_size)
   end
 
   def show
-    institution = extract_institution_id
-    render status: :bad_request, json: { message: 'id must be present' } and return if institution == -1
+    render status: :bad_request, json: { message: 'id must be present' } and return if @institution_id.nil?
 
-    institution = Institution.find_by(id: institution)
+    institution = Institution.find_by(id: @institution_id)
     render status: :bad_request, json: { message: 'school does not exists' } and return if institution.nil?
 
     render status: '200', json: institution
   end
 
-  def extract_institution_id
-    if params['id'].present?
-      params['id'].to_i
-    else
-      -1
-    end
+  private
+
+  def institution_id
+    @institution_id = params.fetch(:id, nil)
   end
 
   def extract_page_presence
-    params['page'].presence || '1'
+    @page = params.fetch(:page, 1)
+    @error_message = 'page must be greater than 0' if @page.to_i < 1
```

Comment:
You can render in the before_action instead of relying on a temp. instance variable
```suggestion
    render status: :bad_request, json: { message: 'page must be greater than 0' } if @page < 1
```